web/api/query.go

```diff
@@ -48,6 +68,12 @@ func (serv MetricsService) Query(w http.ResponseWriter, r *http.Request) {
 	expr := params.Get("expr")
 	asText := params.Get("asText")
 
+	timestamp, err := parseTimestampOrNow(params.Get("timestamp"))
+	if err != nil {
+		http.Error(w, fmt.Sprintf("invalid query timestamp %s", err), http.StatusBadRequest)
```

Comment:
Good point, it should just do `fmt.Fprint(w, ast.ErrorToJSON(err))` like we do below on query errors. There's no content negotiation via Accept-header yet (see TODO below), but for all sane intents and purposes, this should only ever return JSON, and we should remove the text output format here anyways. Will do this later.